docs/proposals/001-in-place-upgrades.md

```diff
@@ -0,0 +1,257 @@
+<!--
+This section MUST reference any work that is out of scope for this proposal.
+Out of scope items are typically unknowns that we do not yet have a clear idea
+of how to solve, so we explicitly do not tackle them until we have more
+information.
+
+This section is very useful to help guide the implementation details section
+below, or serve as reference for future proposals.
+-->
+
+The in-place upgrades only address the upgrades of Canonical Kubernetes and it's respective dependencies. Which means changes on the OS front/image would not be handled since the underlying machine image stays the same. This would be handled by a rolling upgrade as usual.
+
+# Implementation Details
+
+## API Changes
+<!--
+This section MUST mention any changes to the k8sd API, or any additional API
+endpoints (and messages) that are required for this proposal.
+
+Unless there is a particularly strong reason, it is preferable to add new v2/v3
+APIs endpoints instead of breaking the existing APIs, such that API clients are
+not affected.
+-->
+### `POST /x/capi/snap-refresh`
+
+```go
+type SnapRefreshRequest struct {
+	// Channel is the channel to refresh the snap to.
+	Channel string
+	// Revision is the revision number to refresh the snap to.
+	Revision string
+	// LocalPath is the local path to use to refresh the snap.
+	LocalPath string
+}
+```
+
+`POST /x/capi/snap-refresh` performs the in-place upgrade with the given options.
+
+The upgrade can be either done with a `Channel`, `Revision` or a local `*.snap` file provided via `LocalPath`. The value of `LocalPath` should be an absolute path.
+
+This endpoint should use `ValidateCAPIAuthTokenAccessHandler("capi-auth-token")` for authentication.
+
+## Bootstrap Provider Changes
+<!--
+This section MUST mention any changes to the bootstrap provider.
+-->
+
+A machine controller called `MachineReconciler` is added which would perform the in-place upgrade if `k8sd.io/in-place-upgrade-to` annotation is set on the machine.
+
+The controller would use the value of this annotation to make an endpoint call to the `/x/capi/snap-refresh` through `k8sd-proxy`. 
+
+The result of this operation will be communicated back to the user via the `k8sd.io/in-place-upgrade-status` annotation. Values being:
+
+* `in-progress` for an upgrade currently in progress
+* `done` for a successful upgrade
+* `failed` for a failed upgrade
+
+After an upgrade process begins:
+* `k8sd.io/in-place-upgrade-status` annotation on the `Machine` would be added/updated with `in-progress`
+
+After a successfull upgrade:
+* `k8sd.io/in-place-upgrade-to` annotation on the `Machine` would be removed
+* `k8sd.io/in-place-upgrade-current` annotation on the `Machine` would be added/updated with the used `{upgrade-option}`.
+* `k8sd.io/in-place-upgrade-status` annotation on the `Machine` would be added/updated with `done`
+
+After a failed upgrade:
+* `k8sd.io/in-place-upgrade-failure` annotation on the `Machine` would be added/updated with the failure message
```

Comment:
let's make this an `upgrade-status-message`. We might want to have human-friendly messages for other statuses as well (e.g. `in-progress` could have something like `upgrade from ... to ... in progress`)

Reply:
Having a status message in annotations does not feel right to me, maybe we could use events for this so users could see through `kubectl describe`.

Reply:
Is `in-place-upgrade-failure` not a status message - with the limitation that it only is populated on failures?

Reply:
Oh it is for sure, I was wondering if we should move this out of the annotations completely. Sorry seems like I worded it a bit weird 🙂